controller/state.go

```diff
@@ -801,6 +800,51 @@ func (m *appStateManager) CompareAppState(app *v1alpha1.Application, project *v1
 	return &compRes, nil
 }
 
+// useDiffCache will determine if the diff should be calculated based
+// on the existing live state cache or not.
+func useDiffCache(noCache bool, manifestInfos []*apiclient.ManifestResponse, sources []v1alpha1.ApplicationSource, app *v1alpha1.Application, manifestRevisions []string, statusRefreshTimeout time.Duration) bool {
+
+	if noCache {
+		log.Debug("useDiffCache(false): noCache is true")
+		return false
+	}
+	_, refreshRequested := app.IsRefreshRequested()
+	if refreshRequested {
+		log.Debug("useDiffCache(false): refreshRequested")
+		return false
+	}
+	if app.Status.Expired(statusRefreshTimeout) {
+		log.Debug("useDiffCache(false): app.status.expired")
+		return false
+	}
+
+	revisionChanged := len(manifestInfos) != len(sources) ||
```

Comment:
I feel like the `len(manifestInfos) != len(sources)` case would be weird, since `getRepoObjs` should have used `sources` to determine what to get from the repo-server.

Can we detect and log this separately? In case there's a bug, we'll be better able to determine what case is causing it.